gadopt/energy_solver.py

```diff
@@ -53,143 +51,335 @@
+        **kwargs,
+    ) -> None:
+        super().__init__(solution, timestepper, delta_t, **kwargs)
+
+        if isinstance(terms, str):
+            terms = [terms]
+
+        self.terms = terms
+        self.eq_attrs = eq_attrs or {}
+        self.su_diffusivity = su_diffusivity
+
+    def set_equation(self) -> None:
+        if self.su_diffusivity is not None:
+            if (u := self.eq_attrs.get("u")) is None:
+                raise ValueError(
+                    "'u' must be included into `eq_attrs` if `su_diffusivity` is given."
+                )
+
+            self.eq_attrs["su_nubar"] = self.set_su_nubar(u, self.su_diffusivity)
+
+        eq_terms = [self.terms_mapping[term] for term in self.terms]
+
+        self.equation = Equation(
+            self.test,
+            self.solution_space,
+            eq_terms,
+            mass_term=scalar_eq.mass_term,
+            eq_attrs=self.eq_attrs,
+            bcs=self.weak_bcs,
+        )
+
+
+class EnergySolver(GenericTransportBase):
+    """Advances in time the energy conservation equation.
+
+    Note: The solution field is updated in place.
+
+    Arguments:
+      approximation:
+        G-ADOPT approximation defining terms in the system of equations
+      solution:
+        Firedrake function for temperature
+      u:
+        Firedrake function for velocity
+      timestepper:
+        Runge-Kutta time integrator employing an explicit or implicit numerical scheme
+      delta_t:
+        Simulation time step
+      solution_old:
+        Firedrake function holding the solution field at the previous time step
+      bcs:
+        Dictionary specifying boundary conditions (identifier, type, and value)
+      solver_parameters:
+        Dictionary of solver parameters or a string specifying a default configuration
+        provided to PETSc
+      su_diffusivity:
+        Float activating the streamline-upwind stabilisation scheme and specifying the
+        corresponding diffusivity
+
+    """
+
+    def __init__(
+        self,
+        approximation: BaseApproximation,
+        solution: Function,
+        u: Function,
+        /,
+        timestepper: RungeKuttaTimeIntegrator,
+        delta_t: Constant,
+        *,
+        su_diffusivity: float | None = None,
+        **kwargs,
+    ) -> None:
+        super().__init__(solution, timestepper, delta_t, **kwargs)
+
+        self.approximation = approximation
+        self.u = u
+        self.su_diffusivity = su_diffusivity
+
+    def set_equation(self) -> None:
+        rho_cp = self.approximation.rhocp()
+
+        eq_attrs = {
+            "advective_velocity_scaling": rho_cp,
+            "diffusivity": self.approximation.kappa(),
+            "reference_for_diffusion": self.approximation.Tbar,
+            "sink_coeff": self.approximation.linearized_energy_sink(self.u),
+            "source": self.approximation.energy_source(self.u),
+            "u": self.u,
+        }
+
+        if self.su_diffusivity is not None:
```

Comment:
Do you need to repeat this code?

Reply:
I had a go, and I think I can get away without repeating. Thank you.